### PR TITLE
Add LoRAs to the model manager

### PR DESCRIPTION
**Pull request by zopieux**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [x] No, because: 
    - I wasn't aware one had to discuss with the team beforehand to contribute. I did join the Discord however.
    - I tried ["raising my hand"](https://github.com/invoke-ai/InvokeAI#contributing) in #dev-chat on Discord but I do not have permission to write there and could not find instructions on how to get access on the readme or on the server.
    - I need this feature myself anyway and I am upstreaming my contributions in the hope this can be useful to maintainers. I'd rather “ask for forgiveness rather than permission”, especially when donating my time to open-source software.

## Have you updated all relevant documentation?
- [ ] Yes
- [x] No, I do not believe this requires documentation except for a changelog entry maybe.

## Description

Add LoRA models to the manager tab, in addition to checkpoints and diffusers.

## Related Tickets & Documents

I do not believe anyone requested this.

## QA Instructions, Screenshots, Recordings

1. Go to model manager
2. Click on LoRAs (or use the main list)
3. See the model info
4. Delete a LoRA

## Added/updated tests?

- [ ] Yes
- [x] No: there are not tests for `frontend/web`

### CODE REVIEW

**Comment by blessedcoolant** [comment]
The discuss with the InvokeAI team is not a mandatory thing. We just recommend it because there might be a lot of stuff in the works and we do not want you to waste time working on something that someone else might be tackling already. So don't worry too much about it. Feel free to contribute.

Thanks for this PR. I'll take a look at it in sometime and we go ahead from there.

**Comment by zopieux** [comment]
It was reported to me on Discord that the “saving” feature on LoRAs is nonsensical because they do not have persistent state, like main models do, so edits only live as long as the backend lifetime, making this PR quite useless beyond the listing & deleting features. Making it a draft pending more feedback.

**Comment by blessedcoolant** [comment]
Yeah. Functionality wise everything seems to work as advertised. Like you pointed out, coz the LoRA's data is parsed on boot, it is not being stored anywhere so the reboot invalidates any info changes we made.

The question we need to ask here if we wanna do `lora_models.yaml` to track this data as we do with regular models. But this needs to be a lot more hands off than the `models.yaml`.

I'd like to bring in @lstein into the conversation.

One way to get this done properly would be to ---

1. Parse all LoRA models as we do at the moment and store them to a `lora_models.yaml`.
2. The next time we boot, we do the parsing again and if a lora model path is already in the `lora_models.yaml`, then we ignore it and add any new lora models that have been added to the folder.
3. But this mean we'll also need to check for any lora files that have been removed so there aren't any lingering entries in the yaml.
4. Unlike regular models we don't want the user to be constantly adding and deleting LoRA's because there's no conversion needed.

If @lstein can set that up, I think we can retain the edit feature as it is.

**Comment by lstein** [comment]
> Yeah. Functionality wise everything seems to work as advertised. Like you pointed out, coz the LoRA's data is parsed on boot, it is not being stored anywhere so the reboot invalidates any info changes we made.
> 
> The question we need to ask here if we wanna do `lora_models.yaml` to track this data as we do with regular models. But this needs to be a lot more hands off than the `models.yaml`.
> 
> I'd like to bring in @lstein and @StAlKeR7779 into the conversation.
> 
> One way to get this done properly would be to ---
> 
> 1. Parse all LoRA models as we do at the moment and store them to a `lora_models.yaml`.
> 2. The next time we boot, we do the parsing again and if a lora model path is already in the `lora_models.yaml`, then we ignore it and add any new lora models that have been added to the folder.
> 3. But this means we'll also need to check for any lora files that have been removed so there aren't any lingering entries in the yaml.
> 4. Unlike regular models we don't want the user to be constantly adding and deleting LoRA's because there's no conversion needed.
> 
> If @lstein can set that up, I think we can retain the edit feature as it is. I see benefits to it because the user can save trigger words on the LoRA Description and etc.

I like the idea of writing model info to disk. My original model manager design was to have all the model types, including LoRAs, written out to `models.yaml`, and we can still do that just by removing a flag that suppresses all of the non-main model types when syncing memory to disk.

Is there a strong argument for separating out the models among multiple yaml files vs keeping all the persistent model info in a single file?

**Comment by psychedelicious** [comment]
I imagine that would involved abstracting out the model managers current yaml strategy into a generic storage adapter, so when it is initialized we can pass in a custom implementation of the storage adapter.

We *do* have an ItemStorageAbc to provide that abstract base and a corresponding sqlite concrete class. I think the data should be modeled in pydantic if we go this route.

**Comment by zopieux** [comment]
OK this PR was a mess, it didn't actually delete the LoRA when clicking the bin button. 

* Made the edit form readonly with a call-out that it might get eventually supported.
* Added deletion code and tested it removed the files & invalidated the UI.

De-drafting this since I believe it's more useful now as a first step towards better metadata management.

File: invokeai/frontend/web/src/features/ui/components/tabs/ModelManager/subpanels/ModelManagerPanel/ModelListItem.tsx [hunk 1]
```diff
@@ -60,14 +60,21 @@ export default function ModelListItem(props: ModelListItemProps) {
                 title: `${t('modelManager.modelDeleteFailed')}: ${
                   model.model_name
                 }`,
-                status: 'success',
+                status: 'error',
```
Comment (zopieux) [inline invokeai/frontend/web/src/features/ui/components/tabs/ModelManager/subpanels/ModelManagerPanel/ModelListItem.tsx:63]: Fixed this bug I found while testing.

File: invokeai/frontend/web/src/features/ui/components/tabs/ModelManager/subpanels/ModelManagerPanel/ModelListItem.tsx [hunk 2]
```diff
@@ -9,29 +9,26 @@ import { selectIsBusy } from 'features/system/store/systemSelectors';
...
   setSelectedModelId: (v: string | undefined) => void;
 };
 
-const modelBaseTypeMap = {
-  'sd-1': 'SD1',
-  'sd-2': 'SD2',
-  sdxl: 'SDXL',
-  'sdxl-refiner': 'SDXLR',
-};
-
```
Comment (zopieux) [inline invokeai/frontend/web/src/features/ui/components/tabs/ModelManager/subpanels/ModelManagerPanel/ModelListItem.tsx:29]: Refactored this constant as we already have a bunch of these enumerations in `constants.ts`.

**Comment by hipsterusername** [comment]
Going to tag in @lstein to opine here

**Comment by lstein** [comment]
The changes need to be persistent so I'm going to experiment with the quick-and-easy solution of having the LoRAs written out to `models.yaml` for now, and then work towards an `ItemStorageAbc` -style backend. I don't really want people directly editing `models.yaml` anyway.

**Comment by zopieux** [comment]
Note that in its current state this PR allows listing and deletion, and doesn't allow changing the metadata. It's not blocked on a backend change.

**Comment by lstein** [comment]
I've made LoRAs editable (change name, base type, description, and location) and persistent in `models.yaml`. I don't understand typescript very well, so please check this over. It seems to be working.

**Review by lstein — APPROVED** [review]
I like it! Thank you so much for your contribution.

File: invokeai/frontend/web/src/services/api/endpoints/models.ts [hunk 3]
```diff
@@ -52,9 +52,18 @@ type UpdateMainModelArg = {
...
+type UpdateLoRAModelArg = {
+  base_model: BaseModelType;
+  model_name: string;
+  body: LoRAModelConfig;
+};
+
 type UpdateMainModelResponse =
   paths['/api/v1/models/{base_model}/{model_type}/{model_name}']['patch']['responses']['200']['content']['application/json'];
 
+type UpdateLoRAModelResponse =
```
Comment (zopieux) [inline invokeai/frontend/web/src/services/api/endpoints/models.ts:64]: Could you just alias the type above?

**Review by zopieux — COMMENTED** [review]
Thanks for the changes

File: invokeai/frontend/web/src/features/lora/components/ParamLoraSelect.tsx [hunk 4]
```diff
@@ -54,7 +54,7 @@ const ParamLoRASelect = () => {
       });
     });
 
-    return data.sort((a, b) => (a.disabled && !b.disabled ? 1 : -1));
+    return data.sort((a, b) => (a.disabled && !b.disabled ? -1 : 1));
```
Comment (blessedcoolant) [inline invokeai/frontend/web/src/features/lora/components/ParamLoraSelect.tsx:57]: This can be reverted here. I've deployed a more robust fix in #4001

Comment (lstein) [reply invokeai/frontend/web/src/features/lora/components/ParamLoraSelect.tsx]: Thanks. Reverted.

File: invokeai/frontend/web/src/services/api/endpoints/models.ts (hunk 3 above)
Comment (zopieux) [reply invokeai/frontend/web/src/services/api/endpoints/models.ts:64]: Just an optional nit. Literally just assign it to the name above to reduce copy/pasting.

**Review by psychedelicious — APPROVED** [review]
Thanks @zopieux ! Everything here looks great. 

I've rebased on main, resolving a couple minor conflicts, and force-pushed to update the branch.

One thing I noticed in testing is that by allowing users to change model names, we give forfeit stable model identifiers. This is because the model ID is derived from the name.

This causes at least two user-impacting issues:
- If you change the name of a model, your metadata can no longer positively identify the model.
- When you change the name of a model in the Model Manager UI, the model ID also changes and the selected model no longer exists, so it is effectively deselected (actually, there is still a selected model ID internally, but it is no longer valid.

That's not an issue for this PR, but definitely something we should fix. I'll make an issue for broader improvements to how we store and manage this data.

**Comment by psychedelicious** [comment]
I've created #4043 to continue the discussion on improvements to the model management backend

**Pull request merged**